File: bilfind_app/lib/models/user_model.dart
class UserModel {
  UserModel({
    required this.id,
    required this.email,
    required this.name,
    this.profilePhoto,
    required this.familyName,
    required this.departmant,
    required this.createdAt,
    required this.latestStatus,
    required this.favoritePostIds,
    required this.ownPostIds,
    required this.isAdmin,
    this.mailSubscription = true,
  });
  late final String id;
  late final String email;
  late final String name;
  String? profilePhoto;
  late final String familyName;
  late final DateTime createdAt;
  late final String departmant;
  late final String latestStatus;
  late final List<String> favoritePostIds;
  late final List<String> ownPostIds;
  late final bool isAdmin;
  late bool mailSubscription;

  UserModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    email = json['email'];
    name = json['name'];
    profilePhoto = json['profilePhoto'];
    familyName = json['familyName'];
    createdAt = DateTime.parse(json['createdAt']);
    departmant = json['departmant'];
    latestStatus = json['latestStatus'];
    favoritePostIds = List.castFrom<dynamic, String>(json['favoritePostIds']);
    ownPostIds = List.castFrom<dynamic, String>(json['ownPostIds']);
    isAdmin = json['isAdmin'];
    mailSubscription = json['mailSubscription'];
  }
}

File: backend/src/models/user-model.ts
import { Expose, Transform } from "class-transformer";
import { ObjectId } from "mongodb";
import { Mapper } from "../utils/mapper";
import { Departments } from "../utils/enums";

export class User {
  @Transform((value) => value.obj._id.toString())
  @Expose()
  _id?: ObjectId;

  @Expose()
  email: string;

  @Expose()
  hashedPassword: string;

  @Expose()
  name: string;

  @Expose()
  createdAt: Date;

  @Expose()
  profilePhoto?: string;

  @Expose()
  familyName: string;

  @Expose()
  departmant: Departments;

  @Expose()
  latestStatus: UserStatus;

  @Expose()
  favoritePostIds: string[];

  @Expose()
  ownPostIds: string[];

  @Expose()
  ownReportIds: string[];

  @Expose()
  isAdmin: boolean;

  @Expose()
  mailSubscription: boolean;
}

export class UserResponseDTO {
  @Expose()
  id: String;

  @Expose()
  email: string;

  @Expose()
  name: string;

  @Expose()
  profilePhoto?: string;

  @Expose()
  familyName: string;

  @Expose()
  createdAt: Date;

  @Expose()
  departmant: Departments;

  @Expose()
  latestStatus: UserStatus;

  @Expose()
  favoritePostIds: string[];

  @Expose()
  ownPostIds: string[];

  @Expose()
  ownReportIds: string[];

  @Expose()
  isAdmin: boolean;

  @Expose()
  mailSubscription: boolean;
}

export const mapToUserResponseDTO = (user: User): UserResponseDTO => {
  return Mapper.map(UserResponseDTO, {
    ...user,
    id: user._id!.toString(),
  });
};

export enum UserStatus {
  WAITING = "WAITING",
  VERIFIED = "VERIFIED",
  BANNED = "BANNED",
  DELETED = "DELETED",
}
